File: Chapter 02/StatePatternExample.py
class StateExample:
    def do_specificAction(self, context):
        pass
class StateIdle(StateExample):
    def do_specificAction(self, context):
        print("Microplayer is in the Idle state.")
        context.set_state(self)
class StatePlaying(StateExample):
    def do_specificAction(self, context):
        print("Microplayer is in the Playing state.")
        context.set_state(self)
class StatePaused(StateExample):
    def do_specificAction(self, context):
        print("Microplayer is in the Paused state.")
        context.set_state(self)
class Context:
    def __init__(self):
        self._state = None
    def set_state(self, state):
        self._state = state
    def perform_specificAction(self):
        self._state.do_specificAction(self)

if __name__ == "__main__":
    microplayer = Context()

    idle_state = StateIdle()
    playing_state = StatePlaying()
    paused_state = StatePaused()

    microplayer.set_state(idle_state)

    microplayer.perform_action()
    
    microplayer.set_state(playing_state)
    microplayer.perform_specificAction()

    microplayer.set_state(paused_state)
    microplayer.perform_specificAction()

File: Chapter 02/StatePatternExample.java
interface StateExample {
    void doSpecificAction(Context context);
}


class StateIdle implements StateExample {
    public void doSpecificAction(Context context) {
        System.out.println("Microplayer is in the Idle state.");
        context.setState(this);
    }
}
class StatePlaying implements StateExample {
    public void doSpecificAction(Context context) {
        System.out.println("Microplayer is in the Playing state.");
        context.setState(this);
    }
}
class StatePaused implements StateExample {
    public void doSpecificAction(Context context) {
        System.out.println("Microplayer is in the Paused state.");
        context.setState(this);
    }
}


class Context {
    private StateExample state;
    public void setState(StateExample state) {
        this.state = state;
    }
    public void performSpecificAction() {
        state.doSpecificAction(this);
    }
}


public class StatePatternExample {
    public static void main(String[] args) {
        Context microplayer = new Context();
        StateIdle idleState = new StateIdle();
        StatePlaying playingState = new StatePlaying();
        StatePaused pausedState = new StatePaused();

        microplayer.setState(idleState);
        microplayer.performSpecificAction();
        microplayer.setState(playingState);
        microplayer.performSpecificAction();
        
        microplayer.setState(pausedState);
        microplayer.performSpecificAction();
    }
}
